syft/pkg/cataloger/generic/cataloger.go

```diff
@@ -118,20 +118,21 @@ func (c *Cataloger) Catalog(ctx context.Context, resolver file.Resolver) ([]pkg.
 	}
 
 	for _, req := range c.selectFiles(resolver) {
-		location, parser := req.Location, req.Parser
+		log.WithFields("path", req.Location.RealPath).Trace("parsing file contents")
 
-		log.WithFields("path", location.RealPath).Trace("parsing file contents")
+		invokeParser := func(location file.Location, parser Parser) ([]pkg.Package, []artifact.Relationship, error) {
```

Comment:
could this be a static function over a closure? it would be simpler to reason about.

Reply:
Agreed. It will have a lot more parameters (env and logger), but it will be easier to reason about, so that's probably the right trade off. I'll try to push up a rev.